### PR TITLE
Add generation number to data and queries

### PR DESCRIPTION
**Pull request by mqliang**
## Description

Add generation number to data and queries, so when we do comp test during rolling update, we can ingest data with different generation and query against that generation.

A good description should include pointers to an issue or design document, etc.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: pinot-integration-tests/src/test/resources/compat-tests/configs/FeatureTest1-schema.json [hunk 1]
```diff
@@ -78,12 +82,13 @@
       "name": "doubleMetric1"
     }
   ],
-  "timeFieldSpec" : {
-    "incomingGranularitySpec" : {
+  "dateTimeFieldSpecs" : [
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/resources/compat-tests/configs/FeatureTest1-schema.json:85]: Nice, thanks for doing this

File: pinot-integration-tests/src/test/resources/compat-tests/configs/feature-test-1-realtime.json [hunk 2]
```diff
@@ -35,7 +35,7 @@
     "enableDefaultStarTree": false,
     "enableDynamicStarTreeCreation": false,
     "loadMode": "MMAP",
-    "noDictionaryColumns": [],
+    "noDictionaryColumns": ["textDim1"],
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/resources/compat-tests/configs/feature-test-1-realtime.json:38]: You wont need this if you merge after this PR
https://github.com/apache/incubator-pinot/pull/6636

File: pinot-integration-tests/src/test/resources/compat-tests/configs/FeatureTest1-schema.json [hunk 3]
```diff
@@ -1,5 +1,9 @@
 {
   "dimensionFieldSpecs": [
+    {
+      "dataType": "INT",
+      "name": "__GENERATIONNUMBER__"
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/resources/compat-tests/configs/FeatureTest1-schema.json:5]: change the field name to be `generationNumber`

File: pinot-integration-tests/src/test/resources/compat-tests/data/FeatureTest1-data-00.csv [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-s1-6,s2-6,7611,7621,m1-5-0;m1-5-1;m1-5-2,m2-2-0,392;462,6662;782,C++ golang python,deed0507,k1;k2;k3;k4;k5,7;8;9;10;20,"{""k1"":7,""k2"":8,""k3"":9,""k4"":10,""k5"":20}",101,251,262.1,263.10, # 3 values in MV
-s1-6,s2-6,7611,7621,m1-5-0;m1-5-1;m1-5-2,m2-2-0,392;462,6662;782,C++ golang python,deed0507,k1;k2;k3;k4;k5,7;8;9;10;20,"{""k1"":7,""k2"":8,""k3"":9,""k4"":10,""k5"":20}",2147483647,251,262.1,263.10, # MAX_INT in int metric
-s1-6,s2-6,7611,7621,m1-5-0;m1-5-1;m1-5-2,m2-2-0,392;462,6662;782,C++ golang python,deed0507,k1;k2;k3;k4;k5,7;8;9;10;20,"{""k1"":7,""k2"":8,""k3"":9,""k4"":10,""k5"":20}",2147483647,251,262.1,263.10, # MAX_INT in int metric
-s1-7,s2-7,6766,6777,m1-6-0;m1-6-1;m1-6-2;m1-6-3,m2-6-0;m2-6-1,392;462,6662;782,golang Java,d54d0507,k1;k2;k3;k4;k5,31;31;32;32;32,"{""k1"":31,""k2"":31,""k3"":32,""k4"":32,""k5"":32}",87,251,262.10,263.10
+# __GENERATION_NUMBER__ stringDimSV1 stringDimSV2 longDimSV1 longDimSV2 stringDimMV1 stringDimMV2 intDimMV1 intDimMV2 textDim1 mapDim1__KEYS mapDim1__VALUES mapDim2json intMetric1 longMetric1 floatMetric1 doubleMetric1
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/resources/compat-tests/data/FeatureTest1-data-00.csv:2]:
```suggestion
# generationNumber stringDimSV1 stringDimSV2 longDimSV1 longDimSV2 stringDimMV1 stringDimMV2 intDimMV1 intDimMV2 textDim1 mapDim1__KEYS mapDim1__VALUES mapDim2json intMetric1 longMetric1 floatMetric1 doubleMetric1
```

File: pinot-integration-tests/src/test/resources/compat-tests/data/FeatureTest1-data-00.csv [hunk 5]
```diff
@@ -1,13 +1,13 @@
...
+GENERATION_NUMBER,"s1-0",s2-0,1,2,m1-0-0;m1-0-1,m2-0-0;m2-0-1,3;4,6;7,Java C++ Python,01a0bc,k1;k2;k3;k4;k5,1;1;2;2;2,"{""k1"":1,""k2"":1,""k3"":2,""k4"":2,""k5"":2}",10,11,12.1,13.1
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/resources/compat-tests/data/FeatureTest1-data-00.csv:3]:
```suggestion
__GENERATION_NUMBER__,"s1-0",s2-0,1,2,m1-0-0;m1-0-1,m2-0-0;m2-0-1,3;4,6;7,Java C++ Python,01a0bc,k1;k2;k3;k4;k5,1;1;2;2;2,"{""k1"":1,""k2"":1,""k3"":2,""k4"":2,""k5"":2}",10,11,12.1,13.1
```
Same for the other lines as well

File: pinot-integration-tests/src/test/resources/compat-tests/queries/feature-test-1-sql.queries [hunk 6]
```diff
@@ -1,89 +1,89 @@
 # Aggregation
-SELECT count(*) FROM FeatureTest1
-SELECT sum(intMetric1), sumMV(intDimMV1), min(intMetric1), minMV(intDimMV2), max(longDimSV1), maxMV(intDimMV1) FROM FeatureTest1
-SELECT count(longDimSV1), countMV(intDimMV1), avg(floatMetric1), avgMV(intDimMV2), minMaxRange(doubleMetric1), minMaxRangeMV(intDimMV2) FROM FeatureTest1
-SELECT percentile(longDimSV1, 80), percentileMV(intDimMV1, 90), percentileEst(longDimSV1, 80), percentileEstMV(intDimMV1, 90), percentileTDigest(longDimSV1, 80), percentileTDigestMV(intDimMV1, 90) FROM FeatureTest1
-SELECT distinctCount(longDimSV1), distinctCountMV(intDimMV1), distinctCountHLL(longDimSV1), distinctCountHLLMV(intDimMV1) FROM FeatureTest1
+SELECT count(*) FROM FeatureTest1 WHERE __GENERATIONNUMBER__ = GENERATION_NUMBER
```
Comment (mcvsubbu) [inline pinot-integration-tests/src/test/resources/compat-tests/queries/feature-test-1-sql.queries:2]:
```suggestion
SELECT count(*) FROM FeatureTest1 WHERE generationNumber = __GENERATION_NUMBER__
```

Same for the other queries. The column name should be `generationNumber`. The string that you sbstitute can be `__GENERATION+NUMBER__`

**Review by mcvsubbu — COMMENTED** [review]
overall, lgtm

File: pinot-integration-tests/src/test/resources/compat-tests/configs/feature-test-1-realtime.json (hunk 2 above)
Comment (mqliang) [reply pinot-integration-tests/src/test/resources/compat-tests/configs/feature-test-1-realtime.json:38]: Sure, will remove it if #6636 get merged first.

**Review by mcvsubbu — APPROVED** [review]

**Pull request merged**